core/container.go

```diff
@@ -157,18 +157,18 @@ func (container *Container) FS(ctx context.Context) (*Directory, error) {
 	return NewDirectory(ctx, st, "", payload.Platform)
 }
 
-func (container *Container) WithFS(ctx context.Context, st llb.State, platform specs.Platform) (*Container, error) {
+func (container *Container) WithFS(ctx context.Context, dir *Directory, platform specs.Platform) (*Container, error) {
 	payload, err := container.ID.decode()
 	if err != nil {
 		return nil, err
 	}
 
-	stDef, err := st.Marshal(ctx, llb.Platform(platform))
+	dirPayload, err := dir.ID.decode()
 	if err != nil {
 		return nil, err
 	}
 
-	payload.FS = stDef.ToPB()
+	payload.FS = dirPayload.LLB
 	payload.Platform = platform
```

Comment:
Should I ensure that `dirPayload.Platform` == `platform`?

Reply:
I don't think so, and if there's any changes needed in regards to that I'll make them when I rebase the multiplatform PR onto the new API. That one adds tests for these sorts of various issues (and will need more once rebased).